raft.go

```diff
@@ -36,38 +36,72 @@ func (a *Autopilot) NumVoters() (int, error) {
 // its all done will trigger autopilot to remove dead servers if there
 // are any. Servers added by this method will start in a non-voting
 // state and later on autopilot will promote them to voting status
-// if desired by the configured promoter.
+// if desired by the configured promoter. If too many removals would
+// be required that would cause leadership loss then an error is returned
+// instead of performing any Raft configuration changes.
 func (a *Autopilot) AddServer(s *Server) error {
 	cfg, err := a.getRaftConfiguration()
 	if err != nil {
 		a.logger.Error("failed to get raft configuration", "error", err)
 		return err
 	}
 
+	var existingVoter bool
+	var voterRemovals []raft.ServerID
+	var nonVoterRemovals []raft.ServerID
+	var numVoters int
 	for _, server := range cfg.Servers {
+		if server.Suffrage == raft.Voter {
+			numVoters++
+		}
+
 		if server.Address == s.Address && server.ID == s.ID {
 			// nothing to be done as the addr and ID both already match
 			return nil
-		} else if server.Address == s.Address || server.ID == s.ID {
-			// we know that we have a partial match  so we should remove then re-add the server.
-			if err := a.removeServer(server.ID); err != nil {
-				if server.ID == s.ID {
-					return fmt.Errorf("error removing server with duplicate ID %q: %w", server.ID, err)
-				} else {
-					return fmt.Errorf("error removing server with duplicate address %q: %w", server.Address, err)
-				}
+		} else if server.ID == s.ID {
```

Comment:
When the ids match the only thing else in the raft config is the address and suffrage. This branch just wants to detect whether we will need to add the server as a voter or non-voter.

In theory Raft will not demote a voter even if we were to use the AddNonVoter call for both but as that behavior seems counter-intuitive I thought it best to go ahead and provide a code path to call AddVoter instead.